### PR TITLE
Tcp additions

### PR DESCRIPTION
**Pull request by MusicMonkey5555**
Added several little additions to the TCPObject such as sending a file, callbacks for when everything is finished, and extra script methods.

### CODE REVIEW

File: Engine/source/app/net/tcpObject.cpp [hunk 1]
```diff
@@ -400,6 +405,29 @@ void TCPObject::send(const U8 *buffer, U32 len)
...
+	//Read each byte into our buffer
+	Vector<U8> buffer(readFile.getStreamSize());
+	U8 byte;
+	while(readFile.read(&byte))
+	{
+		buffer.push_back(byte);
+	}
+
+	//Send the buffer
+	send(buffer.address(), buffer.size());
```
Comment (jamesu) [inline Engine/source/app/net/tcpObject.cpp]: This is not guaranteed to send all the data, especially with a large file. What you need is a mechanism to send chunks of the file data per tick.

Comment (jamesu) [inline Engine/source/app/net/tcpObject.cpp]: Also `while(readFile.read(&byte))` and pushing each byte to the vector has to be the worst way I can imagine of reading a file. Use `read(numBytes, buffer)` at least. :)

Comment (eightyeight) [inline Engine/source/app/net/tcpObject.cpp]: I remembered you submitted something similar to this James, but didn't have time to check if it was exactly this.

Comment (jamesu) [inline Engine/source/app/net/tcpObject.cpp]: This was https://github.com/GarageGames/Torque3D/pull/658 - however I pulled it due to there not being "a real consensus on this about the implementation". From what I can tell the only similarity between @MusicMonkey5555 's changes (in total) and that branch are the incorporation of the bytesWritten parameter in Net::send.

Comment (MusicMonkey5555) [inline Engine/source/app/net/tcpObject.cpp]: Interesting about the file size. I guess we never hit that limit, but make sense. I wrote a bunch of script code that basically used TCP object to do a bunch of HTTP stuff including opening up a SocketStream to send stuff to and from Google Earth. Most of the files we were sending were pretty small art files or like a JS file so there wasn't an issue. I will look into it and try to improve it, but it works in those limited cases.

And I know there is an HTTP object, but we needed to make rapid changes and the HTTP object didn't seem to:
1. do everything we need
2. had some issues with it working at all

As for the performance issue with reading the file, I'll improve that with your suggestion, thanks!

Comment (MusicMonkey5555) [inline Engine/source/app/net/tcpObject.cpp]: Oh also yeah the bytesWritten part I actually got from you code. My comment on running into an issue when it was queued was on there and I just send the entire file as it worked for what we needed it for. So they are somewhat related. My comments on it still stand though that you run into issues with that implementation of a send queue.

**Comment by MusicMonkey5555** [comment]
After taking a look at the latest changes to the file I noticed all the WebSocket reading support wasn't added, so you could only write. Therefore I added that to this in case someone wants it.

**Comment by Lopuska** [comment]
Good additions, also a very secure change as doesn't modify anything, just adds! Particularly important for network code. Thank you.

**Pull request merged**